File: C++/Arrays/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        
        vector<vector<int>> res;
        
        for(int i = 0; i < n; i++){
            // Removing the duplicate wrt to i
            if(i > 0 && nums[i] == nums[i-1])
                continue;
            
            int l = i + 1;
            int r = n - 1;
            int x = nums[i];
            
            while(l < r){
                if(x + nums[l] + nums[r] == 0){
                    res.push_back(vector<int>{x, nums[l], nums[r]});
                    
                    // Removing the Duplicates wrt l and r
                    while (l < r && nums[l] == nums[l+1]) 
                        l++;
                    
                    while (l < r && nums[r] == nums[r-1]) 
                        r--;
                    
                    l++;
                    r--;
                }
                
                else if(x + nums[l] + nums[r] < 0){
                    l++;
                }
                
                else
                    r--;
            }
        }
        
        return res;
    }
};

File: C++/Arrays/Recvision_2025/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []

        for i in range(len(nums)):
            # Duplicate from the second element onwards
            if i > 0 and nums[i] == nums[i-1]:
                    continue
            
            l = i+1
            r = len(nums)-1

            while l < r:
                sum = nums[i] + nums[l] + nums[r]

                if sum < 0:
                    l += 1
                elif sum > 0:
                    r -= 1
                if sum == 0:
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1
                    # Make sure no duplicate
                    while nums[l] == nums[l-1] and l < r:
                        l += 1
        
        return res
